File: VB/DevExpress.XtraReports.CustomControls/XRTreeListDesigner.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Data
Imports System.Drawing.Design
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.Data.Browsing.Design
Imports DevExpress.Data.Design
Imports DevExpress.XtraReports.Design

Namespace DevExpress.XtraReports.CustomControls
	Public Class XRTreeListDesigner
		Inherits XRDataContainerControlDesigner

		Public Sub New()
			MyBase.New()
		End Sub

		Protected Overrides Sub RegisterActionLists(ByVal list As System.ComponentModel.Design.DesignerActionListCollection)
			MyBase.RegisterActionLists(list)
			list.Add(New XRTreeListDesignerDataActionList(Me, TryCast(Me.Component, XRTreeList)))
		End Sub

		Protected Overrides Function CreateColumnActionList() As XRDataContainerDesignerColumnActionList
			Return New XRTreeListColumnActionList(Me)
		End Function
	End Class

	Public Class XRTreeListDesignerDataActionList
		Inherits XRComponentDesignerActionList

		Private treeList As XRTreeList

		Public Sub New(ByVal componentDesigner As XRComponentDesigner, ByVal treeList As XRTreeList)
			MyBase.New(componentDesigner)
			Me.treeList = treeList
		End Sub
		Protected Overrides Sub FillActionItemCollection(ByVal actionItems As DesignerActionItemCollection)
			AddPropertyItem(actionItems, "KeyFieldName", "KeyFieldName")
			AddPropertyItem(actionItems, "ParentFieldName", "ParentKeyFieldName")
		End Sub

		<Editor(GetType(XRTreeListFieldNameEditor), GetType(UITypeEditor))>
		Public Property KeyFieldName() As String
			Get
				Return treeList.KeyFieldName
			End Get
			Set(ByVal value As String)
				SetPropertyValue("KeyFieldName", value)
			End Set
		End Property

		<Editor(GetType(XRTreeListFieldNameEditor), GetType(UITypeEditor))>
		Public Property ParentKeyFieldName() As String
			Get
				Return treeList.ParentFieldName
			End Get
			Set(ByVal value As String)
				SetPropertyValue("ParentFieldName", value)
			End Set
		End Property
	End Class

	Public Class XRTreeListFieldNameEditor
		Inherits ColumnNameEditor

		Protected Overrides Function GetDataSource(ByVal context As ITypeDescriptorContext) As Object
			Return DirectCast(context.Instance, ICustomDataContainer).GetDataSource()
		End Function
	End Class

	Public Class XRTreeListColumnActionList
		Inherits XRDataContainerDesignerColumnActionList

		Public Sub New(ByVal componentDesigner As XRComponentDesigner)
			MyBase.New(componentDesigner)
		End Sub

		Protected Overrides Sub AddHeadersPropertyItem(ByVal actionItems As DesignerActionItemCollection)
			AddPropertyItem(actionItems, control.FieldHeaderName, "Columns")
		End Sub

		<Editor(GetType(XRCollectionEditor), GetType(UITypeEditor))>
		Public ReadOnly Property Columns() As XRTreeListColumnCollection
			Get
				Return TryCast(MyBase.Headers, XRTreeListColumnCollection)
			End Get
		End Property
	End Class
End Namespace

File: CS/DevExpress.XtraReports.CustomControls/XRTreeListDesigner.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Data;
using System.Drawing.Design;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.Data.Browsing.Design;
using DevExpress.Data.Design;
using DevExpress.XtraReports.Design;

namespace DevExpress.XtraReports.CustomControls
{
    public class XRTreeListDesigner : XRDataContainerControlDesigner
    {
        public XRTreeListDesigner() : base() { }

        protected override void RegisterActionLists(System.ComponentModel.Design.DesignerActionListCollection list)
        {
            base.RegisterActionLists(list);
            list.Add(new XRTreeListDesignerDataActionList(this, this.Component as XRTreeList));
        }

        protected override XRDataContainerDesignerColumnActionList CreateColumnActionList()
        {
            return new XRTreeListColumnActionList(this);
        }
    }

    public class XRTreeListDesignerDataActionList : XRComponentDesignerActionList
    {
        XRTreeList treeList;

        public XRTreeListDesignerDataActionList(XRComponentDesigner componentDesigner, XRTreeList treeList)
            : base(componentDesigner)
        {
            this.treeList = treeList;
        }
        protected override void FillActionItemCollection(DesignerActionItemCollection actionItems)
        {
            AddPropertyItem(actionItems, "KeyFieldName", "KeyFieldName");
            AddPropertyItem(actionItems, "ParentFieldName", "ParentKeyFieldName");
        }

        [
        Editor(typeof(XRTreeListFieldNameEditor), typeof(UITypeEditor))
        ]
        public string KeyFieldName
        {
            get { return treeList.KeyFieldName; }
            set
            {
                SetPropertyValue("KeyFieldName", value);
            }
        }

        [
        Editor(typeof(XRTreeListFieldNameEditor), typeof(UITypeEditor))
        ]
        public string ParentKeyFieldName
        {
            get { return treeList.ParentFieldName; }
            set
            {
                SetPropertyValue("ParentFieldName", value);
            }
        }
    }

    public class XRTreeListFieldNameEditor : ColumnNameEditor
    {
        public override object GetDataSource(ITypeDescriptorContext context)
        {
            return ((ICustomDataContainer)context.Instance).GetDataSource();
        }
    }

    public class XRTreeListColumnActionList : XRDataContainerDesignerColumnActionList
    {
        public XRTreeListColumnActionList(XRComponentDesigner componentDesigner)
            : base(componentDesigner)
        {
        }

        protected override void AddHeadersPropertyItem(DesignerActionItemCollection actionItems)
        {
            AddPropertyItem(actionItems, control.FieldHeaderName, "Columns");
        }

        [Editor(typeof(XRCollectionEditor), typeof(UITypeEditor))]
        public XRTreeListColumnCollection Columns
        {
            get
            {
                return base.Headers as XRTreeListColumnCollection;
            }
        }
    }
}
